Expand AEAD limits to consider multi-user security. (#1)
chris-wood/base-drafts: draft-ietf-quic-tls.md

    date: 2020-02-21
    target: "https://www.felixguenther.info/docs/QUIPS2020_RobustChannels.pdf"

  MultiUserSecurity:
    title: "The Multi-user Security of GCM, Revisited: Tight Bounds for Nonce Randomization"
    author:
      - ins: V. T. Hoang
      - ins: S. Tessaro
      - ins: A. Thiruvengadam
    date: 2018-10-18

>     target: https://dl.acm.org/doi/10.1145/3243734.3243816

Reviewer: You can cite this using `{{?GCM-MU=DOI.10.1145/3243734.3243816}}` inline.

discarded.

## Limits on Key Usage

> The usage limits defined in TLS 1.3 exist for protection against attacks

Reviewer: This still jumps in a bit cold.

```suggestion
This document sets usage limits for AEAD keys to ensure that overuse of a single key does
not give an adversary a disproportionate advantage in attacking the confidentiality
and integrity of communications.

The usage limits defined in TLS 1.3 exist for protection against attacks
```

Or some such nonsense.

Author: I'm happy with that as an intro, though might tweak it a bit to emphasize that limits don't always apply to keys (but to connections).

## Minimum Key Update Frequency

The usage limits defined in TLS 1.3 exist for protection against attacks
on confidentiality and apply to successful applications of AEAD protection. The
integrity protections in authenticated encryption also depend on limiting the
number of attempts to forge packets. TLS achieves this by closing connections
after any record fails an authentication check. In comparison, QUIC ignores any
packet that cannot be authenticated, allowing multiple forgery attempts.

In addition to counting packets sent, endpoints MUST count the number of
received packets that fail authentication during the lifetime of a connection.
If the number of packets sent or fail authentication within the connection
(across all keys) exceeds a limit that is specific to the AEAD in use, the
endpoint MUST immediately close the connection and not attempt to process any
further packets. Applying a limit reduces the probability that an attacker can
distinguish the AEAD in use from a random permutation; see {{AEBounds}},
{{ROBUST}}, and {{?GCM-MU=DOI.10.1145/3243734.3243816}}.

> Note:

Reviewer: I think that we can drop this note now that the failed authentications are counted globally.

Author: Good point -- dropped.

## Minimum Key Update Frequency

Key updates MUST be initiated before usage limits on packet protection keys are
exceeded. For the cipher suites mentioned in this document, the limits in
Section 5.5 of {{!TLS13}} apply. {{!TLS13}} does not specify a limit for
AEAD_AES_128_CCM, but the analysis in {{ccm-bounds}} shows that a limit of 2^23
packets can be used to obtain the same confidentiality protection as the limits

> specified in TLS.

Reviewer: I didn't get enough time to do this today, but what I think we need here is some of the framing from this.  And that likely comes with a new section title and maybe even some restructuring.

Author: I reorganized things a bit. There's a parallel flow for the connection and integrity limit text now that think works. @ekr, what do you think of the new text?

> ## Limits on Key Usage

Reviewer: Like the tweak.  What do you think about the heading?

```suggestion
## Limits on AEAD Usage
```

Author: I originally left it as "Limits on Key Usage" for consistency with RFC8446, though I guess that's not the best rationale. I'm fine with the suggestion.

after any record fails an authentication check. In comparison, QUIC ignores any
packet that cannot be authenticated, allowing multiple forgery attempts.

Endpoints MUST count the number of received packets that fail authentication
for each set of keys and during the lifetime of a connection. If the number of
packets that fail authentication with the same key exceeds a limit that is
specific to the AEAD in use, the endpoint MUST stop using those keys.
Endpoints MUST initiate a key update before reaching this limit.
If a key update is not possible, the endpoint MUST immediately close the
connection. If the number of packets that fail authentication within the
connection (across all keys) exceeds a limit that is specific to the AEAD in

> use, the endpoint MUST immediately close the connection. Applying these limits

Reviewer: An immediate close can still involve sending and receiving packets, so you need to add " and not attempt to process any further packets".  Or something along these lines.

reduces the probability that an attacker is able to successfully forge a
packet in both single- and multi-user security models; see {{AEBounds}},
{{ROBUST}}, and {{MultiUserSecurity}}. 

Note:

> : Due to the way that header protection protects the Key Phase, packets that

Reviewer: This looks like a rewrapping only.  Maybe revert that.

  packet protection keys.  To avoid exhaustion of both sets of keys, it might be
  necessary to initiate two key updates in succession.

Key updates MUST be initiated before usage limits on packet protection keys are
exceeded. For AEAD_AES_128_GCM and AEAD_AES_256_GCM, the confidentiality limit
is 2^27 encrypted packets, and the integrity limit is 2^54 encrypted or forged
packets; see {{?GCM-MU}} and {{gcm-bounds}}. For AEAD_CHACHA20_POLY1305, the
limit on number of packets that fail authentication is 2^93; see {{AEBounds}}.
For AEAD_AES_128_CCM, the confidentiality limit is 2^23 encrypted packets, and
the integrity limit is 2^23.5 forged packets; see {{ccm-bounds}}. Future

> analyses and specifications MAY relax these limits.

Reviewer: We probably need to split this paragraph now.  One for the text on key updates: update before you get to L[confidentiality].  And a separate one for stop using the connection when you get to L[integrity].

Reviewer: How about:

```suggestion
Key updates MUST be initiated before confidentiality limits for the selected
AEAD are exceeded.  For AEAD_AES_128_GCM and AEAD_AES_256_GCM, the
confidentiality limit is 2^27 encrypted packets; see {{gcm-bounds}}.  For
AEAD_CHACHA20_POLY1305, the confidentiality limit is greater than the number of
possible packets (2^62) and so can be disregarded.  For AEAD_AES_128_CCM, the
confidentiality limit is 2^23 encrypted packets; see {{ccm-bounds}}.

A connection MUST be closed if the number of packets that fail authentication
exceed the integrity limit for the selected AEAD. For AEAD_AES_128_GCM, and the
integrity limit is 2^54 encrypted or forged packets; see {{gcm-bounds}}.  For
AEAD_AES_256_GCM, the integrity limit For AEAD_CHACHA20_POLY1305, the limit on
number of packets that fail authentication is 2^36; see {{AEBounds}}.  For
AEAD_AES_128_CCM, the integrity limit is 2^23.5 forged packets; see
{{ccm-bounds}}.

Future analyses and specifications MAY relax confidentiality or integrity limits
for an AEAD.
```

I changed the 2^93 to 2^36 for ChaCha.  I think that the 93 was in error.

  packet protection keys.  To avoid exhaustion of both sets of keys, it might be
  necessary to initiate two key updates in succession.

> A connection MUST be closed if the number of encrypted packets exceeds the

Reviewer: I missed this last time around.

I think that this is unrealistic.  People will make new connections, which have the same properties as new keys.  So we should just recommend key updates.

Author: Hmm, fair point. They are functionally the same thing since we're ignoring the number of users. That said, doesn't it seem inconsistent to drop connections in the case of integrity, but then re-key for confidentiality?

AEAD are exceeded.  For AEAD_AES_128_GCM and AEAD_AES_256_GCM, the
confidentiality limit is 2^27 encrypted packets; see {{gcm-bounds}}.  For
AEAD_CHACHA20_POLY1305, the confidentiality limit is greater than the number of
possible packets (2^62) and so can be disregarded.  For AEAD_AES_128_CCM, the
confidentiality limit is 2^23 encrypted packets; see {{ccm-bounds}}.

A connection MUST be closed if the number of packets that fail authentication
exceed the integrity limit for the selected AEAD. For AEAD_AES_128_GCM, and the
integrity limit is 2^54 encrypted or forged packets; see {{gcm-bounds}}.  For

> AEAD_AES_256_GCM, the integrity limit For AEAD_CHACHA20_POLY1305, the limit on

Reviewer 2: Something is wrong with this text.

Author: Fixed -- thanks!

  Where packets might be larger than 2^14 bytes in length, smaller limits might
  be needed.

Any TLS cipher suite that is specified for use with QUIC MUST define
per-connection limits on the use of the associated AEAD function that
preserves margins for confidentiality and integrity. That is, limits MUST be
specified for the number of packets that can be authenticated and for the
number packets that can fail authentication. Providing a reference to any
analysis upon which values are based - and any assumptions used in that

> analysis - allows limits to be adapted to varying usage conditions.

Reviewer: And maybe revert this change.

Author: Yep, done.

# Analysis of Limits on AEAD_AES_128_CCM Usage {#ccm-bounds}

TLS {{?TLS13}} and {{AEBounds}} do not specify per-key or per-connection
limits on usage for AEAD_AES_128_CCM. However, any AEAD that is used with QUIC
requires limits on use that ensure that both confidentiality and integrity are

> preserved. This section documents that analysis.

Reviewer: Perhaps change to: "This section documents an analysis of per-key limits, which - absent an analysis of multi-user security - also set per-connection limits."

  bound on the number of forged packets that an endpoint can reject before
  updating keys.

The analyses that follow rely on a count of the number of block operations involved
in producing each message. For simplicity, and to match the analysis of other AEAD

> functions in {{AEBounds}}, this analysis assumes a packet length of 2^10 blocks
> and a packet size limit of 2^14.

Reviewer: ```suggestion
functions in {{AEBounds}}, this analysis assumes a packet length of 2^10 blocks;
that is, a packet size limit of 2^14 bytes.
```

by v * l, the number of forgery attempts and the size of each packet (in blocks).

- The amount of offline work done by an attacker does not dominate other factors
in the analysis.

### Confidentiality Limit

For confidentiality, Equation (1) in {{?GCM-MU}} establishes that an attacker gains
a distinguishing advantage between a real and random AEAD algorithm of no more than:

~~~
((q + v) * l)^2 / 2^128
~~~

For a target advantage of 2^-60, which matches that used by {{!TLS13}}, this
results in the relation:

~~~
(q + v) <= 2^28
~~~

Assuming `v = q`, endpoints cannot protect more than 2^27 packets in a single
connection without causing an attacker to gain an larger advantage than the
target of 2^-60.

### Integrity Limit

For integrity, Theorem (4.3) in {{?GCM-MU}} establishes that an attacker gains
an advantage in successfully forging a packet of no more than:

~~~
(1 / 2^(8*n)) + ((2 * v) / 2^2*n) + ((2 * o * v) / 2^(k + n))
        + (n * (v + (v * l)) / 2^k)
~~~

The goal is to limit this advantage to 2^-57, to match the target in
{{?TLS13}}. In this inequality, the last term dominates the rest, so the others

> can be removed without affecting the result. This produces the approximation:

Reviewer: We should probably add a new limit for AEAD_AES_256_GCM.  That is, keep `k` in the following equation and then say 2^54 for 128-bit keys, and then "something much larger" for 256-bit keys, but setting a limit of 2^54 is probably good for that too.  (Honestly, if you have discarded that many packets and are still going, good for you.)

Author: Yeah, let's use the same limit for both, but show (in the appendix) the derivation of the k=256 case. Stand by for an update.

Author: Okay, done!

the associated data and ciphertext. This results in a negligible 1 to 3 block
overestimation of the number of operations.

## Analysis AEAD_AES_128_GCM and AEAD_AES_256_GCM Usage Limits {#gcm-bounds}

{{?GCM-MU}} specify concrete bounds for AEAD_AES_128_GCM and AEAD_AES_256_GCM as
used in TLS 1.3 and QUIC. This section documents this analysis using several
simplifying assumptions:

- The number of ciphertext blocks an attacker uses in forgery attempts is bounded
by v * l, the number of forgery attempts and the size of each packet (in blocks).

- The amount of offline work done by an attacker does not dominate other factors
in the analysis.

> 

Reviewer: ```suggestion

The bounds in {{?GCM-MU}} are tighter and more complete than those used in
{{AEBounds}}, which allows for larger limits than those described in {{?TLS13}}.
```

a distinguishing advantage between a real and random AEAD algorithm of no more than:

~~~
((q + v) * l)^2 / 2^128
~~~

For a target advantage of 2^-60, which matches that used by {{!TLS13}}, this
results in the relation:

~~~
(q + v) <= 2^28
~~~

Assuming `v = q`, endpoints cannot protect more than 2^27 packets in a single
connection without causing an attacker to gain an larger advantage than the
target of 2^-60.

### Integrity Limit

For integrity, Theorem (4.3) in {{?GCM-MU}} establishes that an attacker gains
an advantage in successfully forging a packet of no more than:

~~~
(1 / 2^(8 * n)) + ((2 * v) / 2^2*n) + ((2 * o * v) / 2^(k + n))
        + (n * (v + (v * l)) / 2^k)
~~~

The goal is to limit this advantage to 2^-57, to match the target in
{{?TLS13}}. For AEAD_AES_128_GCM, the fourth term in this inequality dominates
the rest, so the others can be removed without affecting the result. This
produces the following approximation:

~~~
v <= 2^54
~~~

For AEAD_AES_256_GCM, the second and fourth terms dominate the rest, so the
others can be removed without affecting the result. This produces the following
approximation:

~~~
v <= 2^182
~~~

> This is substantially smaller than the limit for AEAD_AES_128_GCM. Thus,

Reviewer: ```suggestion
This is substantially smaller than the limit for AEAD_AES_128_GCM. However,
```

> for consistency, this document uses the larger of the two -- 2^54 -- for both.

Reviewer: ```suggestion
this document recommends that the smaller limit from AEAD_AES_128_GCM also
apply to AEAD_AES_256_GCM.
```

(The only important bit being s/larger/smaller/.)

This is partly consistency (because it is easier to use a single number here) and partly being conservative (because that large number is just unbelievable), and partly because that smaller value is already large enough.

Author: Woof. That one was bad. Good catch!

a distinguishing advantage between a real and random AEAD algorithm of no more than:

~~~
((q + v) * l)^2 / 2^128
~~~

For a target advantage of 2^-60, which matches that used by {{!TLS13}}, this
results in the relation:

~~~
(q + v) <= 2^28
~~~

Assuming `v = q`, endpoints cannot protect more than 2^27 packets in a single
connection without causing an attacker to gain an larger advantage than the
target of 2^-60.

### Integrity Limit

For integrity, Theorem (4.3) in {{?GCM-MU}} establishes that an attacker gains
an advantage in successfully forging a packet of no more than:

~~~
(1 / 2^(8 * n)) + ((2 * v) / 2^2*n) + ((2 * o * v) / 2^(k + n))
        + (n * (v + (v * l)) / 2^k)
~~~

The goal is to limit this advantage to 2^-57, to match the target in
{{?TLS13}}. For AEAD_AES_128_GCM, the fourth term in this inequality dominates
the rest, so the others can be removed without affecting the result. This
produces the following approximation:

~~~
v <= 2^54
~~~

For AEAD_AES_256_GCM, the second and fourth terms dominate the rest, so the
others can be removed without affecting the result. This produces the following
approximation:

~~~
v <= 2^182
~~~

> This is substantially smaller than the limit for AEAD_AES_128_GCM. However,

Reviewer 2: Larger?

Author: Fixed.

- The amount of offline work done by an attacker does not dominate other factors
in the analysis.

The bounds in {{?GCM-MU}} are tighter and more complete than those used in
{{AEBounds}}, which allows for larger limits than those described in {{?TLS13}}.

> ### Confidentiality Limit

Reviewer: ```suggestion

### Confidentiality Limit
```
Just in case.

Author: Already done.
